Fixes #30775 - publishing large cv w/ filters fails Pulp 3 yum

ianballou opened the pull request:
This PR limits the amount of content units that can be copied per Pulp 3 RPM Copy API call.  This limit is being introduced because there seems to be a timeout that can occur if the request body is too large.  The extent of the timeout issue is still being tested.  Currently, the limit is set at 10,000 units.

To do the limiting, I essentially iterate over the original Pulp 3 Copy API data config and copy over unit hrefs to a new Pulp 3 Copy API data config until the limit is reached or the units in the original data config runs out.  Then, the Copy API is called and the process repeats.  I ensure that all repos are included in the call so that dependency solving works.

To test:

1) Sync a few large yum repos that have over 10K units together combined
2) Add those repos to a content view with a filter (still ensuring >10K units) and publish
3) Apply my PR
4) Publish again
5) Check that the same content was copied over
6) Try 1-5 but with over 50K content units if your environment can handle it

jjeffers reviewed (commented):
Tested this with RHEL 7 server rpms repo:
![Screenshot from 2020-09-21 10-52-29](https://user-images.githubusercontent.com/266755/93782476-9f391180-fbf8-11ea-93c7-c723c8aec059.png)

ianballou commented on app/services/katello/pulp3/repository/yum.rb:
```diff
@@ -104,14 +107,77 @@ def multi_copy_units(repo_id_map, dependency_solving)
+          tasks = []
+          unit_amount = 0
+          data.config.each { |repo_config| unit_amount += repo_config[:content].size }
+          if unit_amount > UNIT_LIMIT
```
Yeah, I see the merit in that! I'll give it a try.

ianballou commented on app/services/katello/pulp3/repository/yum.rb:
```diff
@@ -104,14 +107,77 @@ def multi_copy_units(repo_id_map, dependency_solving)
+          data_dup
+        end
+
+        def copy_content_chunked(data)
```
My test for `copy_content_chunked` is pretty rudimentary.  I'm just checking the number of times that `copy_content` is called and checking the args.  Since I can't use actual content units I'm stubbing quite a bit.

I'll look at adding some more tests here.

ianballou commented:
I've updated the tests and refactored the chunked copy skipping for small repos.  The chunked copy logic is pretty rigid so skipping the chunking and using the copy API directly seems like the best route.

My new chunked copy test checks the following:
1) The content units copied is correctly limited
2) The content units copied are in the correct order and with the correct repo pair.  Two scenarios of this are tested:
  a) Source-Dest repo pair content units do not overflow, so the next chunking iteration moves on to the next pair
  b) Source-Dest repo pair content units overflow, so the next chunking iteration copies more units from the same pair

One more test of a large repo with filters would probably be good.  I believe this should be all set otherwise.

ianballou commented:
The failure is related to the current Chromium bug.  Looks like everything else passed.

ianballou commented:
[test katello]

ianballou commented:
[test katello]

ianballou commented:
[test katello]

jjeffers reviewed (approved):
APJ - retested, etc - worked well!

ianballou commented:
[test katello]
